src/network/src/esp/src/TCPServer.cpp

```diff
@@ -0,0 +1,109 @@
+#include "TCPServer.h"
+#include "LockGuard.h"
+#include "NetworkConfig.h"
+#include "SocketFactory.h"
+#include "Task.h"
+#include "lwip/sockets.h"
+
+constexpr uint8_t gs_LOOP_RATE = 10;
+
+static void serverReceiveTask(void* context) {
+
+    if (context != nullptr) {
+        while (true) {
+            static_cast<TCPServer*>(context)->receiveTask();
+            Task::delay(gs_LOOP_RATE);
+        }
+    }
+}
+
+TCPServer::TCPServer(ILogger& logger) :
+    m_serverTask("server_socket", tskIDLE_PRIORITY + 1, serverReceiveTask, this),
+    m_socket(NO_SOCKET),
+    m_logger(logger),
+    m_serverMutex(10) {
+    m_isBusy = false;
+
+    CircularBuff_init(&m_circularBuffer, m_buffer.data(), m_buffer.size());
+
+    m_serverTask.start();
+}
+
+bool TCPServer::send(const uint8_t* data, uint16_t length) {
+    (void)data;
+    (void)length;
+    m_logger.log(LogLevel::Error, "Sending not supported on TCP server");
+    return false;
+}
+
+bool TCPServer::receive(uint8_t* data, uint16_t length) {
+    if (data == nullptr || length > m_buffer.size()) {
+        m_logger.log(LogLevel::Warn, "Invalid parameters for Server receive");
+        return false;
+    }
+
+    LockGuard lock = LockGuard(m_serverMutex);
+    CircularBuff_get(&m_circularBuffer, data, length);
+    return true;
+}
+
+bool TCPServer::isReady() { return m_socket != NO_SOCKET; }
+
+TCPServer::~TCPServer() {
+    if (m_socket != NO_SOCKET) {
+        lwip_close(m_socket);
+    }
+}
+
+void TCPServer::receiveTask() {
+    // Only receive with a valid socket
+    if (m_socket != NO_SOCKET) {
+        int clientfd;
+        sockaddr_in clientAddr;
+        socklen_t addrlen = sizeof(clientAddr);
+        uint8_t buffer[g_MAX_BUFFER_SIZE];
+        int nbytes;
+
+        // lwip_accept is blocking
+        m_logger.log(LogLevel::Info, "Awaiting connection..");
+        clientfd = lwip_accept(m_socket, (sockaddr*)&clientAddr, &addrlen);
+
+        if (clientfd > 0) {
+            m_isBusy = true;
+            nbytes = lwip_recv(clientfd, buffer, sizeof(buffer), 0);
+            if (nbytes > 0) { // client is active
+                do {
+                    if (nbytes > 0) {
+                        LockGuard lock = LockGuard(m_serverMutex);
+                        CircularBuff_put(&m_circularBuffer, buffer, nbytes);
```

Comment:
Do we really need to copy the data to a circular buffer? In HM we just pipe the receive function from the interface directly to the receive() from lwip. That way, we don't have to copy data two times.

Reply:
The goal of the circular buff is to decouple the receiving rate of the stream and the at which the socket refreshes. It will lessen the time between two lwip_accept calls. This issue is less present on the HiveMind because the socket is a client, but the server should always respond as fast as possible.

Reply:
If we have a dispatcher as in HM, the only delay is the deserialisation which I really don't think will be any longer than the memcpy required for putting the data in the circular buffer. And as I said in my previous comment, if you keep the circular buf, you will still need a mechanism to block the read until the buffer has the correct amount of data, so some kind of notification will need to be added here.

 @xgroleau What do you think?

Reply:
I personally prefer just calling receive from lwip, it makes the code much easier to read, it's simple  and saves you CPU cycles.  I would say to keep it simple with read, and change/optimise if we have the need to. Though maybe it was legitimate. Was it currently an issue and you needed to decouple it as of now?

Reply:
Not really sure how this would save CPU cycles since _lwip_accept_ is blocking. Again, the difference between the HiveMind and this is that the socket needs to accept the connection and then can receive bytes. The current implementation design of the deserializer in HiveMind could cause an issue if it then calls a _send_ with the data, which is a blocking call, that would delay the _receive_ call and potentially cause an issue. While I agree that calling directly the _lwip_rcv_ function could work, I believe this approach could cause problems in the future with higher traffic of messages/connections. This is why I didn't follow the same approach as in HiveMind. With the current implementation, I already observe some higher than normal delays between transmissions across different modules. It is an extra security

Also, on a higher level architecture, having the deserializers calling the _receive_ with a certain number of bytes (and similar with the serializers with the _send_) is not the best design if they are directly linked to the transmission. We have allowed a certain maximum size for our packets on every link (UART, SPI, ethernet...) that we don't know if we are really using due to protobuf setting the size of each transmission. I believe that in the future, a layer should be added between the stream and the transmission calls. If not, we could be sending multiples transactions of one byte. Our design should be closer to a 1 message - 1 transaction, except for larger payloads. We

Reply:
Another reason why calling the _lwip_rcv_ function directly could be problematic is when the client needs to send more data than the server is receiving. The socket would be closed before the client has sent everything and the next _receive_ call could come from another client, thus breaking the input stream since a new message would start. We should make sure that we receive all bytes from the client before closing the socket and returning. 

My implementation doesn't fully prevent it but could reduce the occurence since at least we have the confidence that each send will be fully transmitted. Some more work would be needed there to properly prevent the issue of having an interrupted stream, but I don't think this PR is the place for it